File: test_integration/test_ee_timed_scaling_mix.cpp
/*
 * Copyright (c) 2015, 2016, 2017, 2018, 2019, 2020, Intel Corporation
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in
 *       the documentation and/or other materials provided with the
 *       distribution.
 *
 *     * Neither the name of Intel Corporation nor the names of its
 *       contributors may be used to endorse or promote products derived
 *       from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <string.h>
#include <limits.h>
#include <mpi.h>
#include <string>
#include <vector>
#include <memory>

#include "geopm.h"
#include "Exception.hpp"
#include "ModelRegion.hpp"


int main(int argc, char **argv)
{
    int err = 0;
    int comm_rank;
    int comm_size;
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &comm_size);
    MPI_Comm_rank(MPI_COMM_WORLD, &comm_rank);
    bool is_verbose = false;
    if (comm_rank == 0) {
        for (int arg_idx = 1; arg_idx < argc; ++arg_idx) {
            if (strncmp(argv[arg_idx], "--verbose", strlen("--verbose")) == 0 ||
                strncmp(argv[arg_idx], "-v", strlen("-v")) == 0) {
                is_verbose = true;
            }
        }
    }

    std::string test_name = "test_ee_timed_scaling_mix";
    std::unique_ptr<geopm::ModelRegion> ignore_model(geopm::ModelRegion::model_region("spin", 0.075, is_verbose));
    uint64_t ignore_region_id = 0;
    err = geopm_prof_region("ignore", GEOPM_REGION_HINT_IGNORE, &ignore_region_id);
    if (err) {
        throw geopm::Exception(test_name, err, __FILE__, __LINE__);
    }
    int repeat = 100;
    double scaling_factor = 1;
    double timed_factor = 1;
    int num_mix = 5;
    double mix_factor = 1.0 / (num_mix - 1);
    for (int mix_idx = 0; mix_idx != num_mix; ++mix_idx) {
        int timed_idx = num_mix - 1 - mix_idx;
        int scaling_idx = mix_idx;
        double timed_big_o = timed_factor * mix_factor * timed_idx;
        double scaling_big_o = scaling_factor * mix_factor * scaling_idx;
        std::unique_ptr<geopm::ModelRegion> timed_model(geopm::ModelRegion::model_region("timed_scaling-unmarked", timed_big_o, is_verbose));
        std::unique_ptr<geopm::ModelRegion> scaling_model(geopm::ModelRegion::model_region("scaling-unmarked", scaling_big_o, is_verbose));
        char region_name[NAME_MAX];
        region_name[NAME_MAX - 1] = '\0';
        snprintf(region_name, NAME_MAX - 1, "timed-%.2f-scaling-%.2f", timed_big_o, scaling_big_o);
        uint64_t region_id = 0;
        err = geopm_prof_region(region_name, GEOPM_REGION_HINT_UNKNOWN, &region_id);
        if (err) {
            throw geopm::Exception(test_name, err, __FILE__, __LINE__);
        }
        for (int rep_idx = 0; rep_idx != repeat; ++rep_idx) {
            err = geopm_prof_enter(region_id);
            if (err) {
                throw geopm::Exception(test_name, err, __FILE__, __LINE__);
            }
            timed_model->run();
            scaling_model->run();
            err = geopm_prof_exit(region_id);
            if (err) {
                throw geopm::Exception(test_name, err, __FILE__, __LINE__);
            }
            err = geopm_prof_enter(ignore_region_id);
            if (err) {
                throw geopm::Exception(test_name, err, __FILE__, __LINE__);
            }
            ignore_model->run();
            err = geopm_prof_exit(ignore_region_id);
            if (err) {
                throw geopm::Exception(test_name, err, __FILE__, __LINE__);
            }
            MPI_Barrier(MPI_COMM_WORLD);
        }
    }
    MPI_Finalize();
    return err;
}

File: integration/test/test_ee_timed_scaling_mix.py
#!/usr/bin/env python3
#
#  Copyright (c) 2015 - 2021, Intel Corporation
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions
#  are met:
#
#      * Redistributions of source code must retain the above copyright
#        notice, this list of conditions and the following disclaimer.
#
#      * Redistributions in binary form must reproduce the above copyright
#        notice, this list of conditions and the following disclaimer in
#        the documentation and/or other materials provided with the
#        distribution.
#
#      * Neither the name of Intel Corporation nor the names of its
#        contributors may be used to endorse or promote products derived
#        from this software without specific prior written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
#  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
#  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
#  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
#  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
#  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
#  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
#  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
#  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
#  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#

"""Test the energy efficient agent by running a mix of frequency-
   sensitive and frequency-insensitive regions.

"""

from __future__ import absolute_import

import sys
import re
import unittest
import os
import glob

import geopm_context
import geopmpy.io
import geopmpy.agent
import geopmpy.error
import geopm_test_launcher
import util


class AppConf(object):
    """Class that is used by the test launcher as a geopmpy.io.BenchConf
    when running the script as a benchmark.

    """
    def write(self):
        """No configuration files are required.

        """
        pass

    def get_exec_path(self):
        """Path to bencmark

        """
        script_dir = os.path.dirname(os.path.realpath(__file__))
        return os.path.join(script_dir, '.libs', 'test_ee_timed_scaling_mix')

    def get_exec_args(self):
        return []


@util.skip_unless_cpufreq()
@util.skip_unless_optimized()
@unittest.skip('Disabled pending overhaul of agent.')
class TestIntegrationEETimedScalingMix(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        """Create launcher, execute benchmark and set up class variables.

        """
        sys.stdout.write('(' + os.path.basename(__file__).split('.')[0] +
                         '.' + cls.__name__ + ') ...')
        test_name = 'test_ee_timed_scaling_mix'
        cls._report_path = test_name + '.report'
        cls._trace_path = test_name + '.trace'
        cls._skip_launch = not util.do_launch()
        cls._agent_conf_path = test_name + '-agent-config.json'
        if not cls._skip_launch:
            num_node = 2
            num_rank = 2
            min_freq = geopm_test_launcher.geopmread("CPUINFO::FREQ_MIN board 0")
            sticker_freq = geopm_test_launcher.geopmread("CPUINFO::FREQ_STICKER board 0")
            agent_conf = geopmpy.agent.AgentConf(cls._agent_conf_path,
                                                 'energy_efficient',
                                                 {'frequency_min':min_freq,
                                                  'frequency_max':sticker_freq})
            launcher = geopm_test_launcher.TestLauncher(AppConf(),
                                                        agent_conf,
                                                        cls._report_path,
                                                        cls._trace_path,
                                                        time_limit=6000)
            launcher.set_num_node(num_node)
            launcher.set_num_rank(num_rank)
            launcher.run(test_name)


    def test_monotone_frequency(self):
        """Test that agent selects lower frequencies for regions with more
        stream.

        """
        report = geopmpy.io.RawReport(self._report_path)
        host_names = report.host_names()
        mix_name_re = re.compile(r'timed-[0-9]*\.[0-9]*-scaling-[0-9]*\.[0-9]*')
        num_re = re.compile(r'[0-9]*\.[0-9]*')
        for host_name in report.host_names():
            result = []
            for region_name in report.region_names(host_name):
                if mix_name_re.search(region_name):
                    fracs = num_re.findall(region_name)
                    stream_frac = float(fracs[0])
                    dgemm_frac = float(fracs[1])
                    region = report.raw_region(host_name, region_name)
                    self.assertTrue('requested-online-frequency' in region,
                                    msg='Learning for region {} did not complete'.format(region_name))
                    result.append((dgemm_frac, stream_frac, region['requested-online-frequency']))
            self.assertEqual(5, len(result))
            result.sort()
            last_freq = 0.0
            for (dgemm_frac, stream_frac, freq) in result:
                self.assertLessEqual(last_freq, freq,
                                     msg='Chosen frequency decreased with increasing dgemm fraction')
                last_freq = freq
            self.assertNotEqual(result[0][2], result[-1][2],
                                msg='Same frequency chosen for dgemm only and stream only.')

    def test_skip_network_regions(self):
        """Test that agent does not learn for regions declared as communication APIs.

        """
        report = geopmpy.io.RawReport(self._report_path)
        host_names = report.host_names()
        for host_name in report.host_names():
            found_barrier = False
            for region_name in report.region_names(host_name):
                if region_name == 'MPI_Barrier':
                    region = report.raw_region(host_name, region_name)
                    self.assertTrue('requested-online-frequency' not in region)
                    found_barrier = True
            self.assertTrue(found_barrier)


if __name__ == '__main__':
    # Call do_launch to clear non-pyunit command line option
    util.do_launch()
    unittest.main()
